packages/resolvers/default/src/DefaultResolver.js

```diff
@@ -175,18 +175,19 @@ class NodeResolver {
     return (parent ? path.dirname(parent) : '') + ':' + filename;
   }
 
-  resolveFilename(filename: string, dir: string, isURL: ?boolean) {
+  async resolveFilename(filename: string, dir: string, isURL: ?boolean) {
     switch (filename[0]) {
       case '/':
         // Absolute path. Resolve relative to project root.
         return path.resolve(this.options.rootDir, filename.slice(1));
 
       case '~':
         // Tilde path. Resolve relative to nearest node_modules directory,
-        // or the project root - whichever comes first.
+        // the nearest directory with package.json or the project root - whichever comes first.
         while (
-          dir !== this.options.rootDir &&
-          path.basename(path.dirname(dir)) !== 'node_modules'
+          dir !== this.options.projectRoot &&
+          path.basename(path.dirname(dir)) !== 'node_modules' &&
+          !(await fs.exists(path.join(dir, 'package.json')))
```

Comment:
Why would there not be a `package.json` file inside a `node_modules` directory? Seems like this would add a bunch of unnecessary checks.

Reply:
Do you mean that this change isn't needed at all or that this check should be simplified (not check for the package.json) if `dir` contains `node_modules`?

Reply:
Perhaps it could be optimized. For example, it the path contains `node_modules` at all, then just walk up to that point. Otherwise, fall back to looking for a `package.json`. This way, we don't check for `package.json` files at all if inside `node_modules`.